packages/grid/data-grid/src/tests/layout.DataGrid.test.tsx

```diff
@@ -64,6 +68,35 @@ describe('<DataGrid /> - Layout & Warnings', () => {
       expect(rect.width).to.equal(400 - 2);
     });
 
+    it('should throw an error if rows props is being mutated. Error should be thrown due to Object.freeze(rows) on rows prop', () => {
+      const TestDataGrid = () => {
+        const [rows, setRows] = React.useState(baselineProps.rows);
+
+        const addRow = () => {
+          // Error: mutating rows directly
+          rows.push({
+            id: 3,
+            brand: 'Louis Vuitton',
+          });
+          setRows(rows);
+        };
+
+        return (
+          <React.Fragment>
+            <div style={{ width: 300, height: 300 }}>
+              <DataGrid {...baselineProps} />
+            </div>
+            <button onClick={addRow}>Add a new row</button>
+          </React.Fragment>
+        );
+      };
+      render(<TestDataGrid />);
+      const addRowButton = screen.getByText('Add a new row');
+      expect(() => {
+        fireEvent.click(addRowButton);
+      }).to.throw();
```

Comment:
This use-case driven test is still failing. Can't figure out why. Need some direction.

Reply:
The error isn't throw at the click event, but during the `rows.push`. You have to wrap the entire test case in an `expect` to properly catch it. Try this:

```js
  it('should throw an error if rows props is being mutated', () => {
    expect(() => {
      // We don't want to freeze baselineProps.rows
      const rows = [...baselineProps.rows];
      render(
        <div style={{ width: 300, height: 300 }}>
          <DataGrid {...baselineProps} rows={rows} />
        </div>,
      );
      rows.push({ id: 3, brand: 'Louis Vuitton' });
    }).to.throw('Cannot add property 3, object is not extensible');
  });
```

Reply:
Thanks @m4theushw . Also nice catch on `baselineProps.rows` should not be freezed. 👍 

Other thing, I removed the error message string assertion `Cannot add property 3, object is not extensible` as it is different for different browsers.
For Firefox:
```
TypeError: can't define array index property past the end of an array with non-writable length
```

For Safari:
```
TypeError: Attempted to assign to readonly property.
```